File: it-agents/skills/code-architect/references/code-examples/php/Command.php
<?php
// Command — encapsulate a request as an object, support undo.
declare(strict_types=1);

// Receiver holding the state.
final class Counter
{
    public int $value = 0;
}

interface Command
{
    public function execute(): void;
    public function undo(): void;
}

final class AddCommand implements Command
{
    public function __construct(private Counter $counter, private int $by) {}

    public function execute(): void { $this->counter->value += $this->by; }
    public function undo(): void { $this->counter->value -= $this->by; }
}

// Invoker records history to support undo.
final class CommandRunner
{
    /** @var Command[] */
    private array $history = [];

    public function run(Command $command): void
    {
        $command->execute();
        $this->history[] = $command;
    }

    public function undoLast(): void
    {
        $command = array_pop($this->history);
        $command?->undo();
    }
}

// Demo
$counter = new Counter();
$runner = new CommandRunner();
$runner->run(new AddCommand($counter, 5));
$runner->run(new AddCommand($counter, 3));
echo "value={$counter->value}" . PHP_EOL; // 8
$runner->undoLast();
echo "value={$counter->value}" . PHP_EOL; // 5

File: it-agents/skills/code-architect/references/code-examples/python/command.py
"""Command — encapsulate a request as an object, support undo."""

from __future__ import annotations

from abc import ABC, abstractmethod


class Counter:
    """The receiver that actually performs the work."""

    def __init__(self) -> None:
        self.value = 0


class Command(ABC):
    @abstractmethod
    def execute(self) -> None:
        ...

    @abstractmethod
    def undo(self) -> None:
        ...


class AddCommand(Command):
    def __init__(self, counter: Counter, delta: int) -> None:
        self._counter = counter
        self._delta = delta

    def execute(self) -> None:
        self._counter.value += self._delta

    def undo(self) -> None:
        self._counter.value -= self._delta


class Invoker:
    """Runs commands and keeps a history for undo."""

    def __init__(self) -> None:
        self._history: list[Command] = []

    def run(self, command: Command) -> None:
        command.execute()
        self._history.append(command)

    def undo_last(self) -> None:
        if self._history:
            self._history.pop().undo()


if __name__ == "__main__":
    counter = Counter()
    invoker = Invoker()
    invoker.run(AddCommand(counter, 5))
    invoker.run(AddCommand(counter, 3))
    print("after two adds:", counter.value)
    invoker.undo_last()
    print("after undo:    ", counter.value)
